apps/gdalalg_vector_sql.cpp

```diff
@@ -0,0 +1,306 @@
+/******************************************************************************
+ *
+ * Project:  GDAL
+ * Purpose:  "sql" step of "vector pipeline"
+ * Author:   Even Rouault <even dot rouault at spatialys.com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2025, Even Rouault <even dot rouault at spatialys.com>
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#include "gdalalg_vector_sql.h"
+
+#include "gdal_priv.h"
+#include "ogrsf_frmts.h"
+#include "ogrlayerpool.h"
+
+#include <set>
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*           GDALVectorSQLAlgorithm::GDALVectorSQLAlgorithm()           */
+/************************************************************************/
+
+GDALVectorSQLAlgorithm::GDALVectorSQLAlgorithm(bool standaloneStep)
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      standaloneStep)
+{
+    AddArg("sql", 0, _("SQL statement(s)"), &m_sql)
+        .SetPositional()
+        .SetRequired()
+        .SetPackedValuesAllowed(false)
+        .SetReadFromFileAtSyntaxAllowed()
+        .SetMetaVar("<statement>|@<filename>")
+        .SetRemoveSQLCommentsEnabled();
+    AddArg("output-layer", standaloneStep ? 0 : 'l', _("Output layer name(s)"),
+           &m_outputLayer);
+    AddArg("dialect", 0, _("SQL dialect (e.g. OGRSQL, SQLITE)"), &m_dialect);
+}
+
+/************************************************************************/
+/*                   GDALVectorSQLAlgorithmDataset                      */
+/************************************************************************/
+
+namespace
+{
+class GDALVectorSQLAlgorithmDataset final : public GDALDataset
+{
+    GDALDataset *m_poSrcDS = nullptr;
```

Comment:
Is there a way to replace raw pointers with smart ones in this (new) class?

Reply:
I've changed to using a reference instead. But ownership of the source dataset is a bit complicated. It is a broader subject, and I'm not sure we can easily fix that in the GDAL 3.x series (nor we'd have the will to break for GDAL 4)